### PR TITLE
Fix DateToDay

### PR DESCRIPTION
**Pull request by sharansukesh1003**
Fix implementation of Zeller's congruence

refer to https://www.geeksforgeeks.org/zellers-congruence-find-day-date/

[Open in Gitpod](https://gitpod.io/from-referrer/)&nbsp;[know more](https://www.gitpod.io/docs/pull-requests/)

### Describe your change:

- [x] Fix a bug or typo in an existing algorithm?

### Checklist:

- [x] I have read [CONTRIBUTING.md](https://github.com/TheAlgorithms/Javascript/blob/master/CONTRIBUTING.md).
- [x] This pull request is all my own work -- I have not plagiarized.
- [x] I know that pull requests will not be merged if they fail the automated tests.
- [x] This PR only changes one algorithm file. To ease review, please open separate PRs for separate algorithms.
- [x] All new JavaScript files are placed inside an existing directory.
- [x] All filenames should use the UpperCamelCase (PascalCase) style. There should be no spaces in filenames.
      **Example:**`UserProfile.js` is allowed but `userprofile.js`,`Userprofile.js`,`user-Profile.js`,`userProfile.js` are not
- [x] All new algorithms have a URL in its comments that points to Wikipedia or other similar explanation.
- [x] If this pull request resolves one or more open issues then the commit message contains `Fixes: #{$ISSUE_NO}`.

### CODE REVIEW

File: Conversions/DateToDay.js [hunk 1]
```diff
@@ -38,25 +19,23 @@ const daysNameList = { // weeks-day
...
-
 const DateToDay = (date) => {
-  // firstly, check that input is a string or not.
+ // firstly, check that input is a string or not.    
   if (typeof date !== 'string') {
     return new TypeError('Argument is not a string.')
   }
-  // extract the date
-  const [day, month, year] = date.split('/').map((x) => Number(x))
+  //   // extract the date
```
Comment (appgurueu) [inline Conversions/DateToDay.js:27]: Double `//`

File: Conversions/DateToDay.js [hunk 2]
```diff
@@ -38,25 +19,23 @@ const daysNameList = { // weeks-day
...
   }
-  // divide year to century and yearDigit value.
-  const yearDigit = (year % 100)
-  const century = Math.floor(year / 100)
-  // Apply the algorithm shown above
-  const weekDay = Math.abs((day + Math.floor((2.6 * calcMonthList[month]) - 0.2) - (2 * century) + yearDigit + Math.floor(yearDigit / 4) + Math.floor(century / 4)) % 7)
-  // return the weekDay name.
-  return daysNameList[weekDay]
+  // date is resolved based on Zeller's congruence.
+  if (month < 3) { --year; month += 12; }
```
Comment (appgurueu) [inline Conversions/DateToDay.js:34]: I'd use `year--` here

File: Conversions/DateToDay.js (hunk 1 above)
Comment (sharansukesh1003) [reply Conversions/DateToDay.js:27]: Ohhh my bad I will fix that.

File: Conversions/DateToDay.js (hunk 2 above)
Comment (sharansukesh1003) [reply Conversions/DateToDay.js:34]: I will make the change also can you explain me why post decrement would be better in this case ? Just curious.

Comment (sharansukesh1003) [reply Conversions/DateToDay.js]: I have changed it to year-- should I make another pull request?

File: Conversions/DateToDay.js [hunk 3]
```diff
@@ -4,30 +4,11 @@
...
-  7: 5,
-  8: 6,
-  9: 7,
-  10: 8,
-  11: 9,
-  12: 10
-}
-
 // show the week day in a number : Sunday - Saturday => 0 - 6
 const daysNameList = { // weeks-day
```
Comment (appgurueu) [inline Conversions/DateToDay.js:13]: Please use an array here.

**Review by appgurueu — COMMENTED** [review]
Of all proposed PRs I prefer this so far as it gets rid of `calcMonthList` (which is a dict rather than a list)

File: Conversions/DateToDay.js [hunk 4]
```diff
@@ -38,25 +19,23 @@ const daysNameList = { // weeks-day
...
   // extract the date
-  const [day, month, year] = date.split('/').map((x) => Number(x))
+  let [day, month, year] = date.split('/').map((x) => Number(x))
   // check the data are valid or not.
   if (day < 0 || day > 31 || month > 12 || month < 0) {
     return new TypeError('Date is not valid.')
   }
-  // divide year to century and yearDigit value.
-  const yearDigit = (year % 100)
+  // date is resolved based on Zeller's congruence.
```
[the remark this answers is not in the document]
Comment (sharansukesh1003) [reply Conversions/DateToDay.js:33]: Ohhh great input, I will change it.

File: Conversions/DateToDay.js [hunk 5]
```diff
@@ -4,32 +4,13 @@
...
-  8: 6,
-  9: 7,
-  10: 8,
-  11: 9,
-  12: 10
-}
-
 // show the week day in a number : Sunday - Saturday => 0 - 6
-const daysNameList = { // weeks-day
+const daysNameDict = { // weeks-day
```
Comment (appgurueu) [inline Conversions/DateToDay.js:13]: Make this an array please: Keys are `0` to `6`.

Comment (sharansukesh1003) [reply Conversions/DateToDay.js]: Changed it to an array.

**Comment by sharansukesh1003** [comment]
Hi, can you please have a look at my last commit and if everything thing is correct, can you approve my PR I need these points for Hacktober 2022 :P, thanks in advance.

File: Conversions/DateToDay.js [hunk 6]
```diff
@@ -4,59 +4,38 @@
...
-  const yearDigit = (year % 100)
+  // count Jan & Feb as months 13 & 14 of the previous year
+  if (month < 3) { year--; month += 12 }
   const century = Math.floor(year / 100)
-  // Apply the algorithm shown above
-  const weekDay = Math.abs((day + Math.floor((2.6 * calcMonthList[month]) - 0.2) - (2 * century) + yearDigit + Math.floor(yearDigit / 4) + Math.floor(century / 4)) % 7)
-  // return the weekDay name.
-  return daysNameList[weekDay]
+  year %= 100
+  const weekDay = (year + Math.floor(year / 4) + Math.floor(century / 4) - 2 * century + Math.floor((26 * (month + 1)) / 10) + day - 1) % 7
```
Comment (appgurueu) [inline Conversions/DateToDay.js:37]: You don't need the `% 7` here if you apply `% 7` below.

**Comment by appgurueu** [comment]
> Hi, can you please have a look at my last commit and if everything thing is correct, can you approve my PR I need these points for Hacktober 2022 :P, thanks in advance.

I am torn on whether I should merge this or https://github.com/TheAlgorithms/JavaScript/pull/1125/files. Initially this seemed like the better option, but the other PR has improved significantly in response to my review comments and is mostly on par with this one.

File: Conversions/DateToDay.js (hunk 6 above)
Comment (sharansukesh1003) [reply Conversions/DateToDay.js]: Hmmm understood, changed it.

**Comment by sharansukesh1003** [comment]
> > Hi, can you please have a look at my last commit and if everything thing is correct, can you approve my PR I need these points for Hacktober 2022 :P, thanks in advance.
> 
> I am torn on whether I should merge this or #1125. Initially this seemed like the better option, but the other PR has improved significantly in response to my review comments and is mostly on par with this one.

Do mine :P, just kidding. I guess, merge the one that is better in terms of performance. Honestly I got to learn a lot from just this one contribution. If there is any other (good first issue) then maybe assign it to me :).

**Comment by sharansukesh1003** [comment]
If you decide to merge the other PR can you label mine as hacktoberfest-accepted that will give me at least 1 point 😀

**Comment by appgurueu** [comment]
Closing as #1125 has been merged. Sorry. Thank you for your effort.

**Comment by sharansukesh1003** [comment]
> Closing as #1125 has been merged. Sorry. Thank you for your effort.

No problem, I learnt a lot thankyou for all the inputs 😀

**Pull request closed without merging**